<a href="https://colab.research.google.com/github/A-l-E-v/ML-Engineer/blob/main/pandas_homework_task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание
# Тема 2.1. Pandas

**Цель**: научиться использовать библиотеку Pandas для анализа и обработки данных.

Представим, что крупный банк поставил вам задачу исследовать датасет со статистикой о платёжеспособности клиентов.

Описание данных
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

## Задача 1. Загрузка данных и первичный анализ.
https://drive.google.com/file/d/1MJRx330kw9pvJ5DyUUvPm65RxnvNulbW/view

- [ ] Загрузите датафрейм из csv файла, выведите первые 10 строк таблицы.
- [ ] Выведите data.info(). Сколько в датасете строчек? Сколько столбцов?
- [ ] Выведите data.describe(). Все ли в порядке с типами столбцов? Все ли из них нужны?
- [ ] Проверьте, есть ли в данных пропуски? Если есть, то подумайте, как их лучше всего обработать? Можно ли их просто выкинуть?

In [217]:
import pandas as pd

# Указываем правильный разделитель (; вместо запятой)
url = 'https://drive.google.com/uc?export=download&id=1MJRx330kw9pvJ5DyUUvPm65RxnvNulbW'
data = pd.read_csv(url, sep=';')

# Проверяем структуру
print(f"Размер данных: {data.shape} (строки, столбцы)\n")

# 2. Вывод первых 10 строк
print("Первые 10 строк данных:")
display(data.head(10))


Размер данных: (21525, 12) (строки, столбцы)

Первые 10 строк данных:


children  days_employed  dob_years education  education_id  \
0         1   -8184.542837         43    высшее             1   
1         1   -3904.059641         37   среднее             2   
2         0   -5454.719932         34   Среднее             2   
3         3   -4001.004790         33   среднее             2   
4         0  330058.089885         54   среднее             2   
5         0    -898.400256         28    высшее             1   
6         0   -2792.825991         44    высшее             1   
7         0    -148.196182         51   СРЕДНЕЕ             2   
8         2   -6721.969340         36    ВЫСШЕЕ             1   
9         0   -2123.093752         42   среднее             2   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      Ж   сотрудник     0  246259.370269   
1   женат / замужем                 0      Ж   сотрудник     0  108717.613679   
2   женат / замужем                 0      М   сотрудник     0  141509.373728   
3   женат / замужем                 0      М   сотрудник     0  259599.693820   
4  гражданский брак                 1      Ж   пенсионер     0  153857.595534   
5  гражданский брак                 1      М   компаньон     0  248090.658456   
6   женат / замужем                 0      Ж   компаньон     0  233310.192763   
7   женат / замужем                 0      М   сотрудник     0  131749.216171   
8  гражданский брак                 1      Ж   сотрудник     0   92981.127451   
9   женат / замужем                 0      М   сотрудник     0  140093.160129   

        purpose  
0  недвижимость  
1    автомобиль  
2  недвижимость  
3   образование  
4       свадьба  
5  недвижимость  
6  недвижимость  
7   образование  
8       свадьба  
9  недвижимость

In [218]:
# 3. Вывод информации о датафрейме
print("\nИнформация о данных (data.info()):")
data.info()

# Точное количество строк и столбцов
n_rows, n_cols = data.shape
print(f"Датафрейм содержит {n_rows} строк и {n_cols} столбцов")



Информация о данных (data.info()):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
Датафрейм содержит 21525 строк и 12 столбцов


In [219]:
# 4. Вывод статистики
print("\nОписательная статистика (data.describe()):")
display(data.describe(include='all'))

# Через атрибуты DataFrame
print(f"Строки: {len(data)}")  # или data.shape[0]
print(f"Столбцы: {len(data.columns)}")  # или data.shape[1]

# Явный вывод
print(f"\nРазмерность данных: {data.shape} (строки, столбцы)")

# Подробный анализ
print("\nПодробный подсчёт:")
print(f"- Всего записей: {data.index.size}")
print(f"- Перечень столбцов ({len(data.columns)}):")
for i, col in enumerate(data.columns, 1):
    print(f"  {i}. {col}")



Описательная статистика (data.describe()):


children  days_employed     dob_years education  education_id  \
count   21525.000000   19351.000000  21525.000000     21525  21525.000000   
unique           NaN            NaN           NaN        15           NaN   
top              NaN            NaN           NaN   среднее           NaN   
freq             NaN            NaN           NaN     13750           NaN   
mean        0.538908   61155.102732     44.293380       NaN      1.817236   
std         1.381587  136602.492615     12.574584       NaN      0.548138   
min        -1.000000  -17837.281404      1.000000       NaN      1.000000   
25%         0.000000   -2665.000917     34.000000       NaN      2.000000   
50%         0.000000   -1167.268443     43.000000       NaN      2.000000   
75%         1.000000    -282.363075     54.000000       NaN      2.000000   
max        20.000000  389702.738461     76.000000       NaN      5.000000   

          family_status  family_status_id gender income_type          debt  \
count             21525      21525.000000  21525       21525  21525.000000   
unique                5               NaN      3           8           NaN   
top     женат / замужем               NaN      Ж   сотрудник           NaN   
freq              12380               NaN  14236       11119           NaN   
mean                NaN          0.972544    NaN         NaN      0.080883   
std                 NaN          1.420324    NaN         NaN      0.272661   
min                 NaN          0.000000    NaN         NaN      0.000000   
25%                 NaN          0.000000    NaN         NaN      0.000000   
50%                 NaN          0.000000    NaN         NaN      0.000000   
75%                 NaN          1.000000    NaN         NaN      0.000000   
max                 NaN          4.000000    NaN         NaN      1.000000   

        total_income       purpose  
count   1.935100e+04         21525  
unique           NaN             4  
top              NaN  недвижимость  
freq             NaN         10840  
mean    1.623996e+05           NaN  
std     9.988242e+04           NaN  
min     2.004725e+04           NaN  
25%     9.996156e+04           NaN  
50%     1.406674e+05           NaN  
75%     1.973320e+05           NaN  
max     2.197636e+06           NaN

Строки: 21525
Столбцы: 12

Размерность данных: (21525, 12) (строки, столбцы)

Подробный подсчёт:
- Всего записей: 21525
- Перечень столбцов (12):
  1. children
  2. days_employed
  3. dob_years
  4. education
  5. education_id
  6. family_status
  7. family_status_id
  8. gender
  9. income_type
  10. debt
  11. total_income
  12. purpose


In [220]:
# Проверяем отрицательные значения
neg_days = data[data['days_employed'] < 0]
print(f"\nНайдено {len(neg_days)} записей с отрицательным стажем ({len(neg_days)/len(data):.1%})")



# Анализ типов и статистик
print("\n=== Анализ данных ===")
print("1. Типы данных:")
print(data.dtypes)

print("\n3. Пропущенные значения:")
missing = data.isna().sum()
display(missing[missing > 0])

# Анализ аномалий
print("\n4. Выявленные проблемы:")
print(f"- Отрицательный стаж: {len(neg_days)} записей")
print(f"- Максимальное значение стажа: {data['days_employed'].max():,.0f} дней ({data['days_employed'].max()/365:.1f} лет)")
print(f"- Аномалии в детях: {data['children'].value_counts()}")


Найдено 15906 записей с отрицательным стажем (73.9%)

=== Анализ данных ===
1. Типы данных:
children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

3. Пропущенные значения:


days_employed    2174
total_income     2174
dtype: int64


4. Выявленные проблемы:
- Отрицательный стаж: 15906 записей
- Максимальное значение стажа: 389,703 дней (1067.7 лет)
- Аномалии в детях: children
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: count, dtype: int64


In [221]:
print("\n=== Рекомендации ===")
print("1. Для days_employed:")
print("   - Заменить отрицательные значения на модуль: data['days_employed'] = abs(data['days_employed'])")
print("   - Проверить аномально большие значения (>50 лет стажа)")

print("\n2. Для пропусков:")
print("   - Заполнить медианой для числовых столбцов:")
print("     data['days_employed'].fillna(data['days_employed'].median(), inplace=True)")
print("     data['total_income'].fillna(data['total_income'].median(), inplace=True)")

print("\n3. Для категориальных данных:")
print("   - Привести education к нижнему регистру: data['education'] = data['education'].str.lower()")
print("   - Проверить уникальные значения family_status")


=== Рекомендации ===
1. Для days_employed:
   - Заменить отрицательные значения на модуль: data['days_employed'] = abs(data['days_employed'])
   - Проверить аномально большие значения (>50 лет стажа)

2. Для пропусков:
   - Заполнить медианой для числовых столбцов:
     data['days_employed'].fillna(data['days_employed'].median(), inplace=True)
     data['total_income'].fillna(data['total_income'].median(), inplace=True)

3. Для категориальных данных:
   - Привести education к нижнему регистру: data['education'] = data['education'].str.lower()
   - Проверить уникальные значения family_status


## Задача 2. Предобработка данных
- [ ] В некоторых столбцах есть пропущенные значения. Вам необходимо выделить их и заполнить их какими то значениями. Опишите какие пропущенные значения вы обнаружили. Какую долю они составляют в каждом столбце?
- [ ] Определите какой тип данных в столбце total_income. Замените его на более подходящий, используя astype(). Проверьте типы данных в других столбцах на правильность.
- [ ] Значения в столбцах могут быть записаны некорректно. Это называется  артефактами или аномалиями, - битые данные появившиеся по какой-то причине. Это нормально, если вы работаете с реальными данными. Найдите и обработайте аномалии и попробуйте предположить - по какой причине они могли возникнуть.
- [ ] Проверьте данные на дубликаты. При их нахождении удалите их. Предположите причины их появления. Обратите внимание: дубликаты могут быть неявными

In [222]:
## 1. Обработка пропусков
print("=== Анализ пропусков ===")
missing_values = data.isna().sum()
missing_percent = (missing_values / len(data)) * 100
missing_report = pd.DataFrame({'Пропуски': missing_values, 
                              'Доля (%)': missing_percent.round(2)})
display(missing_report[missing_report['Пропуски'] > 0])



=== Анализ пропусков ===


Пропуски  Доля (%)
days_employed      2174      10.1
total_income       2174      10.1

In [223]:
# Заполнение пропусков
data['days_employed'] = data['days_employed'].fillna(data['days_employed'].median())
data['total_income'] = data['total_income'].fillna(data['total_income'].median())

## Коррекция типов данных
print("\n=== Коррекция типов данных ===")
print("Тип total_income до преобразования:", data['total_income'].dtype)
data['total_income'] = data['total_income'].astype('int64')
print("Тип total_income после преобразования:", data['total_income'].dtype)

# Проверка других столбцов
data['children'] = data['children'].astype('int64')
data['dob_years'] = data['dob_years'].astype('int64')
data['education_id'] = data['education_id'].astype('int64')
data['family_status_id'] = data['family_status_id'].astype('int64')




=== Коррекция типов данных ===
Тип total_income до преобразования: float64
Тип total_income после преобразования: int64


In [224]:
print("\n=== Обработка аномалий ===")

# Отрицательные значения стажа
print(f"Найдено {len(data[data['days_employed'] < 0])} отрицательных значений стажа")
data['days_employed'] = data['days_employed'].abs()
# Проверка аномальных значений в столбце 'days_employed'
print("\nАномальные значения в 'days_employed':")
print(data[data['days_employed'] > 365*100])  # Например, значения больше 100 лет

# Предположим, что значения записаны в часах, преобразуем их в дни
data['days_employed'] = data['days_employed'].apply(lambda x: x / 24 if x > 365*100 else x)


# Нереальное количество детей
print("Распределение значений в столбце children:")
print(data['children'].value_counts())
data = data[data['children'] <= 5]  # Удаляем записи с 20 детьми

# Нормализация образования
data['education'] = data['education'].str.lower()


=== Обработка аномалий ===
Найдено 18080 отрицательных значений стажа

Аномальные значения в 'days_employed':
       children  days_employed  dob_years education  education_id  \
4             0  330058.089885         54   среднее             2   
18            0  388272.702805         54   среднее             2   
24            1  328395.394324         58   среднее             2   
25            0  352642.034668         68   среднее             2   
30            1  325514.218460         63   среднее             2   
...         ...            ...        ...       ...           ...   
21505         0  328737.720414         54   среднее             2   
21508         0  374902.782655         63   среднее             2   
21509         0  351296.222500         60    высшее             1   
21518         0  362775.839512         60   СРЕДНЕЕ             2   
21521         0  333619.282008         68   среднее             2   

               family_status  family_status_id gender income

In [225]:
print("\n=== Обработка дубликатов ===")
duplicates_count = data.duplicated().sum()
print(f"Найдено {duplicates_count} полных дубликатов")
data = data.drop_duplicates()

# Проверка неявных дубликатов
print("\nПроверка неявных дубликатов по ключевым полям:")
key_columns = ['dob_years', 'education', 'family_status', 'gender', 'income_type']
duplicates_key = data.duplicated(subset=key_columns).sum()
print(f"Найдено {duplicates_key} потенциальных неявных дубликатов")


=== Обработка дубликатов ===
Найдено 405 полных дубликатов

Проверка неявных дубликатов по ключевым полям:
Найдено 18051 потенциальных неявных дубликатов


In [226]:
print("\nИтоговый размер данных после обработки:", data.shape)


Итоговый размер данных после обработки: (21044, 12)


## Задача 3. Объединение таблиц
- [ ] Создайте новые датафреймы:
    - Первый, где каждому уникальному значению из education соответствует уникальное id_education.
    - Второй - каждому уникальному значению из family_status - id_fam_status
- [ ] Удалите из исходного датафрейма столбцы education и family_status
- [ ] Объедините три полученных таблицы

In [227]:
# 1. Создание новых датафреймов с уникальными значениями и id
# Датафрейм для education
education_df = data[['education']].drop_duplicates().reset_index(drop=True)
education_df['id_education'] = education_df.index

# Датафрейм для family_status
family_status_df = data[['family_status']].drop_duplicates().reset_index(drop=True)
family_status_df['id_fam_status'] = family_status_df.index

In [228]:
# 2. Удаление столбцов education и family_status из исходного датафрейма
data = data.drop(columns=['education', 'family_status'])


In [229]:
# 3. Объединение таблиц
# Объединение с education_df
data = data.merge(education_df, left_on='education_id', right_on='id_education', how='left')

# Объединение с family_status_df
data = data.merge(family_status_df, left_on='family_status_id', right_on='id_fam_status', how='left')

# Вывод первых 10 строк таблицы после объединения
print("\nПервые 10 строк таблицы после объединения:")
print(data.head(10))


Первые 10 строк таблицы после объединения:
   children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8184.542837         43             1                 0      Ж   
1         1    3904.059641         37             2                 0      Ж   
2         0    5454.719932         34             2                 0      М   
3         3    4001.004790         33             2                 0      М   
4         0   13752.420412         54             2                 1      Ж   
5         0     898.400256         28             1                 1      М   
6         0    2792.825991         44             1                 0      Ж   
7         0     148.196182         51             2                 0      М   
8         2    6721.969340         36             1                 1      Ж   
9         0    2123.093752         42             2                 0      М   

  income_type  debt  total_income       purpose            education  \
0  

## Задача 4. Вычисления
- [ ] С помощью отдельной функции присвойте каждому кредитозаемщику категорию соответственно его доходу. Для этого создайте столбец person_total_income c категориями:
    - 0 - 20000 - “E”;
    - 20000 - 45000 - “D”;
    - 45000 - 70000 - “С”;
    - 70000 - 180000 - “B”;
    - 180000 - выше - “A”
- [ ] Разбейте исходные данные на два датафрейма. В первый -- данные только о мужчинах, во второй – о женщинах. Сбросьте индексы в обеих таблицах. Рассчитайте средние значения признаков для обеих таблиц. Что можно о них сказать?
- [ ] Вернемся к исходному большому датасету. Сгруппируйте данные по следующим полям и посчитайте метрики (общее количество людей в группе, количество должников в группе, доля должников в группе):
    - person_total_income
    - children
    - purpose
    - family_status
- [ ] Создайте  вспомогательный столбец have_children, в который внесем информацию есть ли дети у заемщика (true/false)
- [ ] С помощью pivot_table() постройте сводную таблицу, где по вертикали будет family_status, а по горизонтали have_children


Ответьте на следующие целевые вопросы:  
- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?  

In [230]:
# Функция для присвоения категории дохода
def categorize_income(income):
    if income <= 20000:
        return 'E'
    elif income <= 45000:
        return 'D'
    elif income <= 70000:
        return 'C'
    elif income <= 180000:
        return 'B'
    else:
        return 'A'

# Создание столбца person_total_income
data['person_total_income'] = data['total_income'].apply(categorize_income)

# Вывод первых 10 строк таблицы после добавления категорий
print("\nПервые 10 строк таблицы после добавления категорий:")
print(data.head(10))


Первые 10 строк таблицы после добавления категорий:
   children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8184.542837         43             1                 0      Ж   
1         1    3904.059641         37             2                 0      Ж   
2         0    5454.719932         34             2                 0      М   
3         3    4001.004790         33             2                 0      М   
4         0   13752.420412         54             2                 1      Ж   
5         0     898.400256         28             1                 1      М   
6         0    2792.825991         44             1                 0      Ж   
7         0     148.196182         51             2                 0      М   
8         2    6721.969340         36             1                 1      Ж   
9         0    2123.093752         42             2                 0      М   

  income_type  debt  total_income       purpose            educati

In [231]:
# Разделение данных на мужчин и женщин
male_data = data[data['gender'] == 'М'].reset_index(drop=True)
female_data = data[data['gender'] == 'Ж'].reset_index(drop=True)

# Расчет средних значений признаков
male_means = male_data.mean(numeric_only=True)
female_means = female_data.mean(numeric_only=True)

print("\nСредние значения признаков для мужчин:")
print(male_means)

print("\nСредние значения признаков для женщин:")
print(female_means)


Средние значения признаков для мужчин:
children                 0.522328
days_employed         3070.499019
dob_years               41.965532
education_id             1.837566
family_status_id         0.900223
debt                     0.103545
total_income        183544.320123
id_education             1.835800
id_fam_status            0.900223
dtype: float64

Средние значения признаков для женщин:
children                 0.449016
days_employed         4824.363357
dob_years               45.375586
education_id             1.804641
family_status_id         1.024069
debt                     0.071269
total_income        148715.003099
id_education             1.804180
id_fam_status            1.024069
dtype: float64


 Анализ:
 - Средний доход мужчин выше, чем у женщин
 - У мужчин в среднем больше детей


In [232]:
# Группировка данных и расчет метрик
grouped_data = data.groupby(['person_total_income', 'children', 'purpose', 'family_status']).agg(
    total_count=('gender', 'count'),
    debtors_count=('debt', 'sum')
).reset_index()

grouped_data['debtors_share'] = grouped_data['debtors_count'] / grouped_data['total_count']

print("\nГруппированные данные с метриками:")
print(grouped_data.head(10))


Группированные данные с метриками:
  person_total_income  children       purpose          family_status  \
0                   A        -1    автомобиль       гражданский брак   
1                   A        -1  недвижимость         вдовец / вдова   
2                   A        -1  недвижимость        женат / замужем   
3                   A        -1   образование        женат / замужем   
4                   A         0    автомобиль  Не женат / не замужем   
5                   A         0    автомобиль              в разводе   
6                   A         0    автомобиль         вдовец / вдова   
7                   A         0    автомобиль       гражданский брак   
8                   A         0    автомобиль        женат / замужем   
9                   A         0  недвижимость  Не женат / не замужем   

   total_count  debtors_count  debtors_share  
0            1              0       0.000000  
1            1              0       0.000000  
2            4              0 

In [233]:
# Создание вспомогательного столбца have_children
data['have_children'] = data['children'] > 0

print("\nПервые 10 строк таблицы после добавления столбца have_children:")
print(data.head(10))


Первые 10 строк таблицы после добавления столбца have_children:
   children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8184.542837         43             1                 0      Ж   
1         1    3904.059641         37             2                 0      Ж   
2         0    5454.719932         34             2                 0      М   
3         3    4001.004790         33             2                 0      М   
4         0   13752.420412         54             2                 1      Ж   
5         0     898.400256         28             1                 1      М   
6         0    2792.825991         44             1                 0      Ж   
7         0     148.196182         51             2                 0      М   
8         2    6721.969340         36             1                 1      Ж   
9         0    2123.093752         42             2                 0      М   

  income_type  debt  total_income       purpose       

In [234]:
# Построение сводной таблицы
pivot_table = data.pivot_table(index='family_status', columns='have_children', aggfunc='size', fill_value=0)

print("\nСводная таблица:")
print(pivot_table)


Сводная таблица:
have_children          False  True 
family_status                      
Не женат / не замужем   2241    534
в разводе                786    405
вдовец / вдова           835    104
гражданский брак        2713   1399
женат / замужем         7301   4726


Зависимость между наличием детей и возвратом кредита в срок

Для анализа этой зависимости можно использовать столбец have_children и столбец debt.

In [235]:
# Группировка данных по наличию детей и расчет доли должников
children_debt_share = data.groupby('have_children').agg(
    total_count=('debt', 'count'),
    debtors_count=('debt', 'sum')
).reset_index()

children_debt_share['debtors_share'] = children_debt_share['debtors_count'] / children_debt_share['total_count']

print("\nЗависимость между наличием детей и возвратом кредита в срок:")
print(children_debt_share)


Зависимость между наличием детей и возвратом кредита в срок:
   have_children  total_count  debtors_count  debtors_share
0          False        13876           1062       0.076535
1           True         7168            669       0.093331


Зависимость между семейным положением и возвратом кредита в срок

Для анализа этой зависимости можно использовать столбец family_status и столбец debt.

In [236]:
# Группировка данных по семейному положению и расчет доли должников
family_status_debt_share = data.groupby('family_status').agg(
    total_count=('debt', 'count'),
    debtors_count=('debt', 'sum')
).reset_index()

family_status_debt_share['debtors_share'] = family_status_debt_share['debtors_count'] / family_status_debt_share['total_count']

print("\nЗависимость между семейным положением и возвратом кредита в срок:")
print(family_status_debt_share)



Зависимость между семейным положением и возвратом кредита в срок:
           family_status  total_count  debtors_count  debtors_share
0  Не женат / не замужем         2775            273       0.098378
1              в разводе         1191             84       0.070529
2         вдовец / вдова          939             63       0.067093
3       гражданский брак         4112            385       0.093628
4        женат / замужем        12027            926       0.076993


Зависимость между уровнем дохода и возвратом кредита в срок

Для анализа этой зависимости можно использовать столбец person_total_income и столбец debt.

In [237]:
# Группировка данных по уровню дохода и расчет доли должников
income_debt_share = data.groupby('person_total_income').agg(
    total_count=('debt', 'count'),
    debtors_count=('debt', 'sum')
).reset_index()

income_debt_share['debtors_share'] = income_debt_share['debtors_count'] / income_debt_share['total_count']

print("\nЗависимость между уровнем дохода и возвратом кредита в срок:")
print(income_debt_share)



Зависимость между уровнем дохода и возвратом кредита в срок:
  person_total_income  total_count  debtors_count  debtors_share
0                   A         6028            443       0.073490
1                   B        13385           1176       0.087860
2                   C         1383             95       0.068691
3                   D          248             17       0.068548


Как разные цели кредита влияют на его возврат в срок

Для анализа этой зависимости можно использовать столбец purpose и столбец debt.



In [238]:
# Группировка данных по цели кредита и расчет доли должников
purpose_debt_share = data.groupby('purpose').agg(
    total_count=('debt', 'count'),
    debtors_count=('debt', 'sum')
).reset_index()

purpose_debt_share['debtors_share'] = purpose_debt_share['debtors_count'] / purpose_debt_share['total_count']

print("\nЗависимость между целью кредита и возвратом кредита в срок:")
print(purpose_debt_share)


Зависимость между целью кредита и возвратом кредита в срок:
        purpose  total_count  debtors_count  debtors_share
0    автомобиль         4256            400       0.093985
1  недвижимость        10542            779       0.073895
2   образование         3949            369       0.093441
3       свадьба         2297            183       0.079669


# ОБЩИЕ ВЫВОДЫ:

На основе предоставленных результатов можно сделать следующие выводы:

### 1. Зависимость между наличием детей и возвратом кредита в срок:
- **Выводы**:
  - Доля должников среди клиентов с детьми (`have_children = True`) составляет 9.33%, что выше, чем среди клиентов без детей (`have_children = False`), где доля должников составляет 7.65%.
  - Это может указывать на то, что наличие детей увеличивает риск невозврата кредита в срок. Возможно, это связано с дополнительными финансовыми обязательствами, которые возникают у родителей.

### 2. Зависимость между семейным положением и возвратом кредита в срок:
- **Выводы**:
  - Наименьшая доля должников наблюдается среди вдовцов/вдов (6.71%) и разведенных (7.05%).
  - Наибольшая доля должников наблюдается среди неженатых/незамужних (9.84%) и тех, кто находится в гражданском браке (9.36%).
  - Женатые/замужние клиенты имеют долю должников 7.70%, что также ниже, чем у неженатых/незамужних.
  - Это может указывать на то, что семейное положение влияет на финансовую стабильность и, соответственно, на вероятность возврата кредита в срок.

### 3. Зависимость между уровнем дохода и возвратом кредита в срок:
- **Выводы**:
  - Наименьшая доля должников наблюдается среди клиентов с уровнем дохода категории C (6.87%) и D (6.85%).
  - Наибольшая доля должников наблюдается среди клиентов с уровнем дохода категории B (8.79%) и A (7.35%).
  - Это может указывать на то, что клиенты с более высоким доходом (категории A и B) имеют больше финансовых обязательств, что увеличивает риск невозврата кредита в срок.

### 4. Как разные цели кредита влияют на его возврат в срок:
- **Выводы**:
  - Наименьшая доля должников наблюдается среди клиентов, взявших кредит на недвижимость (7.39%).
  - Наибольшая доля должников наблюдается среди клиентов, взявших кредит на автомобиль (9.40%) и свадьбу (7.97%).
  - Это может указывать на то, что кредиты на автомобиль и свадьбу имеют более высокий риск невозврата в срок по сравнению с кредитами на недвижимость и образование.

### Общие выводы:
- Наличие детей, семейное положение, уровень дохода и цель кредита оказывают значительное влияние на вероятность возврата кредита в срок.
- Банки и кредитные организации могут использовать эти выводы для улучшения своих кредитных скоринговых моделей и снижения риска невозврата кредитов.

**Результат:** решение для каждого задания предоставляется в виде программного кода на языке Python и пояснения при наличии вопросов в задании

**Критерии оценивания:**

К1 - загружен датафрейм из csv файла, выведены первые 10 строк таблицы (2 балла)

К2 - выведен data.info(), приведен верный анализ полученной информации (2 балла)

К3 - выведен data.describe(), приведен верный анализ полученной информации (2 балла)

К4 -  верно определено количество пропусков содержащихся в данных, какую долю от общего количества данных они составляют, и есть ли между ними связь. Правильно заполнены пустые значения (2 балла)

К5 - тип данных в столбце  total_income заменен на более подходящий, есть пояснение (2 балла)

К6 - аномалии выявлены и обработаны верно (2 балла)

К7 - дубликаты удалены, описаны возможные причины появления дубликатов в данных (2 балла)

К8 - данные категоризированы, есть объяснение принципа категоризации данных (2 балла)

К9 - даны ответы на целевые вопросы (4 балла)


Максимальное количество баллов —  20 баллов.

Минимальное количество баллов, чтобы преподаватель смог зачесть вашу работу — 10 баллов.